source/MaterialXGraphEditor/Graph.cpp

```diff
@@ -3566,97 +3566,133 @@ void Graph::addNodePopup(bool cursor)
-                            _addNewNode = true;
-                            memset(input, '\0', sizeof(input));
+                            std::string val = getUserNodeDefName(nodeName);
+                            if (ImGui::MenuItem(val.c_str()) || (ImGui::IsItemFocused() && ImGui::IsKeyPressedMap(ImGuiKey_Enter)))
+                            {
+                                addNode(it->second[i]->getNodeString(), val, it->second[i]->getType());
+                                _addNewNode = true;
+                                memset(input, '\0', sizeof(input));
+                            }
                         }
                     }
                 }
-            }
-            else
-            {
-                ImGui::SetNextWindowSizeConstraints(ImVec2(100, 10), ImVec2(-1, 300));
-                if (ImGui::BeginMenu(it->first.c_str()))
+                else
                 {
-                    ImGui::SetWindowFontScale(_fontScale);
-                    for (size_t j = 0; j < it->second.size(); j++)
+                    ImGui::SetNextWindowSizeConstraints(ImVec2(100, 10), ImVec2(-1, 300));
+                    if (ImGui::BeginMenu(it->first.c_str()))
                     {
-                        std::string name = it->second[j][0];
-                        if (ImGui::MenuItem(getUserNodeDefName(name).c_str()) || (ImGui::IsItemFocused() && ImGui::IsKeyPressedMap(ImGuiKey_Enter)))
+                        ImGui::SetWindowFontScale(_fontScale);
+                        for (size_t i = 0; i < it->second.size(); i++)
                         {
-                            addNode(it->second[j][2], getUserNodeDefName(name), it->second[j][1]);
-                            _addNewNode = true;
+                            std::string name = it->second[i]->getName();
+                            std::string val = getUserNodeDefName(name);
+                            if (ImGui::MenuItem(val.c_str()) || (ImGui::IsItemFocused() && ImGui::IsKeyPressedMap(ImGuiKey_Enter)))
+                            {
+                                addNode(it->second[i]->getNodeString(), val, it->second[i]->getType());
+                                _addNewNode = true;
+                            }
                         }
+                        ImGui::EndMenu();
                     }
-                    ImGui::EndMenu();
                 }
             }
-        }
-
-        // Filter nodedefs and add to menu if matches filter
-        for (std::unordered_map<std::string, std::vector<mx::NodeDefPtr>>::iterator it = _nodesToAdd.begin(); it != _nodesToAdd.end(); ++it)
-        {
-            // Filter out list of nodes
-            if (subs.size() > 0)
+            else
             {
-                ImGui::SetNextWindowSizeConstraints(ImVec2(250.0f, 300.0f), ImVec2(-1.0f, 500.0f));
-                for (size_t i = 0; i < it->second.size(); i++)
+                auto it = _extraNodes.find(key);
```

Comment:
I didn't like reusing `auto it` here, looking for some advice on a better structure, changing the name is also possible but felt kind of sloppy/lazy to me

Reply:
Since the original `it` variable is known to be invalid in this code path, perhaps you can just reuse the variable itself, and omit the `auto` keyword from this line?

Reply:
I wanted to do that originally, but the types weren't the same so it didn't allow it, I've refactored the code to consolidate _nodesToAdd and _extraNodes so I think it's better now